File: apps/api/src/auth/auth.service.ts
import {
  BadRequestException,
  ForbiddenException,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { User, UserRole, UserStatus } from '@prisma/client';
import { createHash } from 'node:crypto';
import { nanoid } from 'nanoid';
import { PrismaService } from '../prisma/prisma.service';
import { AuditService } from '../audit/audit.service';
import { MailService } from '../mail/mail.service';
import { PasswordService } from './password.service';
import { TokenService } from './token.service';
import { TwoFactorService } from './two-factor.service';

const VERIFICATION_TTL_MS = 24 * 60 * 60 * 1000; // 24h
const PASSWORD_RESET_TTL_MS = 60 * 60 * 1000; // 1h

export interface RequestContext {
  ip?: string;
  userAgent?: string;
}

export interface LoginResult {
  user: User;
  accessToken: string;
  refreshToken: string;
}

@Injectable()
export class AuthService {
  private readonly maxAttempts: number;
  private readonly lockoutMinutes: number;
  private readonly attemptWindowMinutes: number;
  private readonly webOrigin: string;

  constructor(
    private readonly prisma: PrismaService,
    private readonly passwords: PasswordService,
    private readonly tokens: TokenService,
    private readonly twoFactor: TwoFactorService,
    private readonly mail: MailService,
    private readonly audit: AuditService,
    config: ConfigService,
  ) {
    this.maxAttempts = Number(config.get('AUTH_LOCKOUT_MAX_ATTEMPTS') ?? 5);
    this.lockoutMinutes = Number(
      config.get('AUTH_LOCKOUT_DURATION_MIN') ?? 15,
    );
    this.attemptWindowMinutes = Number(
      config.get('AUTH_LOCKOUT_WINDOW_MIN') ?? 15,
    );
    this.webOrigin =
      config.get<string>('WEB_ORIGIN') ?? 'http://localhost:3000';
  }

  // ──────────────────────────── REGISTER ────────────────────────────

  async register(
    email: string,
    password: string,
    profile: { firstName?: string; lastName?: string },
    ctx: RequestContext,
  ): Promise<{ userId: string }> {
    const normalizedEmail = email.toLowerCase().trim();
    const existing = await this.prisma.user.findUnique({
      where: { email: normalizedEmail },
    });
    if (existing) {
      // Don't leak existence — return a generic 400.
      throw new BadRequestException(
        'If this email is available, you will receive a verification message',
      );
    }

    const passwordHash = await this.passwords.hash(password);
    const user = await this.prisma.user.create({
      data: {
        email: normalizedEmail,
        passwordHash,
        role: UserRole.MEMBER,
        profile: {
          create: {
            firstName: profile.firstName,
            lastName: profile.lastName,
          },
        },
      },
    });

    await this.issueAndSendVerification(user);
    await this.audit.record({
      actorId: user.id,
      action: 'auth.register',
      ip: ctx.ip,
      userAgent: ctx.userAgent,
    });

    return { userId: user.id };
  }

  // ──────────────────────────── LOGIN ────────────────────────────

  async login(
    email: string,
    password: string,
    totpCode: string | undefined,
    expectedRole: UserRole | null,
    ctx: RequestContext,
  ): Promise<LoginResult> {
    const normalizedEmail = email.toLowerCase().trim();
    const user = await this.prisma.user.findUnique({
      where: { email: normalizedEmail },
    });

    // Constant-ish response — same error for missing/bad creds.
    if (!user || user.status !== UserStatus.ACTIVE) {
      await this.audit.record({
        action: 'auth.login.fail',
        ip: ctx.ip,
        userAgent: ctx.userAgent,
        metadata: { email: normalizedEmail, reason: 'unknown_user' },
      });
      throw new UnauthorizedException('Invalid credentials');
    }

    if (user.lockedUntil && user.lockedUntil > new Date()) {
      await this.audit.record({
        actorId: user.id,
        action: 'auth.login.locked',
        ip: ctx.ip,
        userAgent: ctx.userAgent,
      });
      throw new UnauthorizedException(
        'Account temporarily locked. Try again later.',
      );
    }

    const passwordOk = await this.passwords.verify(
      user.passwordHash,
      password,
    );
    if (!passwordOk) {
      await this.recordFailedAttempt(user, ctx);
      throw new UnauthorizedException('Invalid credentials');
    }

    if (expectedRole && user.role !== expectedRole) {
      // E.g. member trying admin login or vice versa.
      await this.audit.record({
        actorId: user.id,
        action: 'auth.login.wrong_portal',
        ip: ctx.ip,
        userAgent: ctx.userAgent,
      });
      throw new ForbiddenException('Wrong sign-in portal for this account');
    }

    // 2FA enforcement.
    if (user.totpEnabled) {
      if (!totpCode) {
        throw new UnauthorizedException({
          message: '2FA code required',
          requires2fa: true,
        });
      }
      const ok = this.twoFactor.verify(totpCode, user.totpSecret!);
      if (!ok) {
        await this.recordFailedAttempt(user, ctx);
        throw new UnauthorizedException('Invalid 2FA code');
      }
    }

    // Success — reset counters, stamp last login.
    const updated = await this.prisma.user.update({
      where: { id: user.id },
      data: {
        failedLoginCount: 0,
        lockedUntil: null,
        lastLoginAt: new Date(),
      },
    });

    const accessToken = this.tokens.signAccessToken(updated);
    const refreshToken = await this.tokens.issueRefreshToken(
      updated.id,
      ctx.ip,
      ctx.userAgent,
    );

    await this.audit.record({
      actorId: user.id,
      action: 'auth.login.success',
      ip: ctx.ip,
      userAgent: ctx.userAgent,
    });

    return { user: updated, accessToken, refreshToken };
  }

  private async recordFailedAttempt(user: User, ctx: RequestContext) {
    const now = new Date();
    const windowStart = new Date(
      now.getTime() - this.attemptWindowMinutes * 60 * 1000,
    );
    // Reset the counter if either (a) the previous failure run aged out of the
    // window, or (b) a prior lockout has now expired (start them fresh).
    const lockoutJustExpired = !!user.lockedUntil && user.lockedUntil < now;
    const counterStale = user.updatedAt < windowStart;
    const shouldReset = lockoutJustExpired || counterStale;

    const nextCount = shouldReset ? 1 : user.failedLoginCount + 1;
    const shouldLock = nextCount >= this.maxAttempts;
    const lockedUntil = shouldLock
      ? new Date(now.getTime() + this.lockoutMinutes * 60 * 1000)
      : lockoutJustExpired
        ? null
        : user.lockedUntil;

    await this.prisma.user.update({
      where: { id: user.id },
      data: { failedLoginCount: nextCount, lockedUntil },
    });

    await this.audit.record({
      actorId: user.id,
      action: shouldLock ? 'auth.login.lockout' : 'auth.login.fail',
      ip: ctx.ip,
      userAgent: ctx.userAgent,
      metadata: { failedCount: nextCount },
    });
  }

  // ──────────────────────────── REFRESH / LOGOUT ────────────────────────────

  refresh(token: string, ctx: RequestContext) {
    return this.tokens.rotate(token, ctx.ip, ctx.userAgent);
  }

  async logout(refreshToken: string | undefined, userId: string | undefined) {
    if (refreshToken) await this.tokens.revokeRefreshToken(refreshToken);
    if (userId) {
      await this.audit.record({ actorId: userId, action: 'auth.logout' });
    }
  }

  // ──────────────────────────── EMAIL VERIFICATION ────────────────────────────

  private async issueAndSendVerification(user: User): Promise<void> {
    const raw = nanoid(48);
    const tokenHash = sha256(raw);
    await this.prisma.emailVerification.create({
      data: {
        userId: user.id,
        tokenHash,
        expiresAt: new Date(Date.now() + VERIFICATION_TTL_MS),
      },
    });
    const link = `${this.webOrigin}/verify-email?token=${raw}`;
    await this.mail.sendEmailVerification(user.email, link);
  }

  async verifyEmail(token: string, ctx: RequestContext): Promise<void> {
    const tokenHash = sha256(token);
    const record = await this.prisma.emailVerification.findUnique({
      where: { tokenHash },
      include: { user: true },
    });
    if (!record || record.usedAt || record.expiresAt < new Date()) {
      throw new BadRequestException('Invalid or expired verification token');
    }
    await this.prisma.$transaction([
      this.prisma.emailVerification.update({
        where: { id: record.id },
        data: { usedAt: new Date() },
      }),
      this.prisma.user.update({
        where: { id: record.userId },
        data: { emailVerifiedAt: new Date() },
      }),
    ]);
    await this.audit.record({
      actorId: record.userId,
      action: 'auth.email.verified',
      ip: ctx.ip,
      userAgent: ctx.userAgent,
    });
  }

  async resendVerification(email: string): Promise<void> {
    const user = await this.prisma.user.findUnique({
      where: { email: email.toLowerCase().trim() },
    });
    if (user && !user.emailVerifiedAt) {
      await this.issueAndSendVerification(user);
    }
    // Always succeed silently to avoid email-enumeration.
  }

  // ──────────────────────────── PASSWORD RESET ────────────────────────────

  async forgotPassword(email: string, ctx: RequestContext): Promise<void> {
    const user = await this.prisma.user.findUnique({
      where: { email: email.toLowerCase().trim() },
    });
    if (user) {
      const raw = nanoid(48);
      const tokenHash = sha256(raw);
      await this.prisma.passwordReset.create({
        data: {
          userId: user.id,
          tokenHash,
          expiresAt: new Date(Date.now() + PASSWORD_RESET_TTL_MS),
        },
      });
      const link = `${this.webOrigin}/reset-password?token=${raw}`;
      await this.mail.sendPasswordReset(user.email, link);
      await this.audit.record({
        actorId: user.id,
        action: 'auth.password.reset_requested',
        ip: ctx.ip,
        userAgent: ctx.userAgent,
      });
    }
    // Always succeed silently.
  }

  async resetPassword(
    token: string,
    newPassword: string,
    ctx: RequestContext,
  ): Promise<void> {
    const tokenHash = sha256(token);
    const record = await this.prisma.passwordReset.findUnique({
      where: { tokenHash },
    });
    if (!record || record.usedAt || record.expiresAt < new Date()) {
      throw new BadRequestException('Invalid or expired reset token');
    }

    const passwordHash = await this.passwords.hash(newPassword);
    await this.prisma.$transaction([
      this.prisma.passwordReset.update({
        where: { id: record.id },
        data: { usedAt: new Date() },
      }),
      this.prisma.user.update({
        where: { id: record.userId },
        data: { passwordHash, failedLoginCount: 0, lockedUntil: null },
      }),
      // Invalidate all sessions on password reset.
      this.prisma.refreshToken.updateMany({
        where: { userId: record.userId, revokedAt: null },
        data: { revokedAt: new Date() },
      }),
    ]);
    await this.audit.record({
      actorId: record.userId,
      action: 'auth.password.reset',
      ip: ctx.ip,
      userAgent: ctx.userAgent,
    });
  }

  // ──────────────────────────── 2FA ────────────────────────────

  async beginTwoFactorSetup(userId: string) {
    const user = await this.prisma.user.findUniqueOrThrow({
      where: { id: userId },
    });
    const setup = await this.twoFactor.generateSetup(user.email);
    // Stash secret unconfirmed; only flip totpEnabled once verified.
    await this.prisma.user.update({
      where: { id: userId },
      data: { totpSecret: setup.secret, totpEnabled: false },
    });
    return setup;
  }

  async confirmTwoFactorSetup(
    userId: string,
    code: string,
    ctx: RequestContext,
  ): Promise<void> {
    const user = await this.prisma.user.findUniqueOrThrow({
      where: { id: userId },
    });
    if (!user.totpSecret) {
      throw new BadRequestException('No 2FA setup in progress');
    }
    const ok = this.twoFactor.verify(code, user.totpSecret);
    if (!ok) throw new BadRequestException('Invalid 2FA code');

    await this.prisma.user.update({
      where: { id: userId },
      data: { totpEnabled: true },
    });
    await this.audit.record({
      actorId: userId,
      action: 'auth.2fa.enabled',
      ip: ctx.ip,
      userAgent: ctx.userAgent,
    });
  }

  async disableTwoFactor(
    userId: string,
    password: string,
    code: string,
    ctx: RequestContext,
  ): Promise<void> {
    const user = await this.prisma.user.findUniqueOrThrow({
      where: { id: userId },
    });
    if (user.role === UserRole.ADMIN) {
      throw new ForbiddenException('Admins cannot disable 2FA');
    }
    const passwordOk = await this.passwords.verify(
      user.passwordHash,
      password,
    );
    if (!passwordOk) throw new UnauthorizedException('Invalid password');
    if (!user.totpEnabled || !user.totpSecret) {
      throw new BadRequestException('2FA is not enabled');
    }
    const codeOk = this.twoFactor.verify(code, user.totpSecret);
    if (!codeOk) throw new BadRequestException('Invalid 2FA code');

    await this.prisma.user.update({
      where: { id: userId },
      data: { totpEnabled: false, totpSecret: null },
    });
    await this.audit.record({
      actorId: userId,
      action: 'auth.2fa.disabled',
      ip: ctx.ip,
      userAgent: ctx.userAgent,
    });
  }

  // ──────────────────────────── ME ────────────────────────────

  async me(userId: string) {
    const user = await this.prisma.user.findUniqueOrThrow({
      where: { id: userId },
      include: { profile: true },
    });
    return {
      id: user.id,
      email: user.email,
      role: user.role,
      status: user.status,
      emailVerified: user.emailVerifiedAt !== null,
      totpEnabled: user.totpEnabled,
      profile: user.profile,
      createdAt: user.createdAt,
    };
  }
}

function sha256(input: string): string {
  return createHash('sha256').update(input).digest('hex');
}

File: apps/api/app/services/auth_service.py
"""
Auth business logic — port of `apps/api/src/auth/auth.service.ts`.

Each public method maps 1:1 to a Nest service method. Behavior parity rules:

  * Lockout: 5 attempts / 15-min window / 15-min lock (env-tunable).
    Reset rule mirrors Nest's: counter resets when EITHER the previous failure
    aged out of the window OR a prior lockout has expired.
  * Email-enumeration safety: register & forgot-password & resend-verification
    all return 'ok' regardless of whether the email exists.
  * Password reset invalidates ALL active refresh tokens for the user (forces
    re-login on every device).
  * Audit writes use fire-and-forget via FastAPI BackgroundTasks (caller-owned).
"""

from __future__ import annotations

import hashlib
import secrets
import string
from datetime import datetime, timedelta, timezone
from typing import Any

from fastapi import BackgroundTasks
from sqlalchemy import func, select, update
from sqlalchemy.orm import joinedload
from sqlalchemy.ext.asyncio import AsyncSession

from ..core import security
from ..core.logging import get_logger
from ..core.security import AuthenticatedUser
from ..core.settings import get_settings
from ..db.models import (
    EmailVerification,
    MemberProfile,
    PasswordReset,
    RefreshToken,
    User,
)
from ..integrations.mail import get_mail_service
from . import audit_service
from . import token_service

log = get_logger("auth")

VERIFICATION_TTL = timedelta(hours=24)
PASSWORD_RESET_TTL = timedelta(hours=1)

# cuid format matches Prisma's @default(cuid())
_CUID_ALPHABET = string.ascii_lowercase + string.digits


def _mint_cuid() -> str:
    return "c" + "".join(secrets.choice(_CUID_ALPHABET) for _ in range(24))


def _utcnow_naive() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def _to_naive_utc(dt: datetime | None) -> datetime | None:
    if dt is None:
        return None
    if dt.tzinfo is not None:
        return dt.astimezone(timezone.utc).replace(tzinfo=None)
    return dt


def _sha256(s: str) -> str:
    return hashlib.sha256(s.encode("utf-8")).hexdigest()


def _normalize_email(e: str) -> str:
    return e.strip().lower()


class AuthError(Exception):
    """Internal — raised by services, mapped to HTTP errors in the router."""

    def __init__(self, code: int, message: str, extra: dict[str, Any] | None = None):
        self.code = code
        self.message = message
        self.extra = extra or {}
        super().__init__(message)


# ───────────── REGISTER ──────────────────────────────────────────────────────


async def register(
    db: AsyncSession,
    *,
    email: str,
    password: str,
    firstName: str | None,
    lastName: str | None,
    ip: str | None,
    userAgent: str | None,
    background: BackgroundTasks,
) -> str:
    normalized = _normalize_email(email)
    existing = (await db.execute(
        select(User).where(User.email == normalized)
    )).scalar_one_or_none()
    if existing is not None:
        # Don't leak existence — same generic 400 NestJS returns.
        raise AuthError(
            400,
            "If this email is available, you will receive a verification message",
        )

    password_hash = await security.async_hash_password(password)
    user_id = _mint_cuid()
    now = _utcnow_naive()
    db.add(
        User(
            id_=user_id,
            email=normalized,
            passwordHash=password_hash,
            role="MEMBER",
            status="ACTIVE",
            createdAt=now,
            updatedAt=now,
        )
    )
    # Flush before the dependent insert — the SQLAlchemy models don't carry
    # ForeignKey() metadata (the generator emits scalar columns only), so the
    # unit-of-work would otherwise re-order these inserts and trip the
    # member_profiles_userId_fkey constraint.
    await db.flush()
    db.add(
        MemberProfile(
            id_=_mint_cuid(),
            userId=user_id,
            firstName=firstName,
            lastName=lastName,
            createdAt=now,
            updatedAt=now,
        )
    )
    await db.flush()

    # Persist the verification token WITH the user, then commit. The account is
    # now durably created regardless of what the mailer does.
    raw = await _issue_verification_token(db, user_id=user_id)

    await db.commit()

    # Send the verification email AFTER commit, best-effort. A mail failure
    # (provider outage, missing key, transient error) must never roll back or
    # fail a completed signup — the token is already stored, so the user can
    # request a resend. Runs as a background task so it never blocks the
    # response either.
    background.add_task(_send_verification_email_safe, normalized, raw)
    background.add_task(
        audit_service.record,
        actorId=user_id,
        action="auth.register",
        ip=ip,
        userAgent=userAgent,
    )
    return user_id


async def _issue_verification_token(db: AsyncSession, *, user_id: str) -> str:
    """Insert an email-verification token row and return its raw (unhashed)
    value for embedding in the verification link. Does NOT send anything."""
    raw = _nanoid(48)
    token_hash = _sha256(raw)
    now = _utcnow_naive()
    db.add(
        EmailVerification(
            id_=_mint_cuid(),
            userId=user_id,
            tokenHash=token_hash,
            expiresAt=now + VERIFICATION_TTL,
            createdAt=now,
        )
    )
    await db.flush()
    return raw


async def _send_verification_email_safe(email: str, raw_token: str) -> None:
    """Best-effort verification email — swallows and logs any failure so it can
    never break a signup that already committed."""
    try:
        s = get_settings()
        link = f"{s.WEB_ORIGIN}/verify-email?token={raw_token}"
        await get_mail_service().send_email_verification(email, link)
    except Exception as e:  # noqa: BLE001
        log.warning("verification_email_send_failed", email=email, error=str(e))


def _nanoid(length: int) -> str:
    # Match nanoid's default URL-safe alphabet (A-Za-z0-9_-) and length.
    alphabet = string.ascii_letters + string.digits + "_-"
    return "".join(secrets.choice(alphabet) for _ in range(length))


# ───────────── LOGIN ─────────────────────────────────────────────────────────


async def login(
    db: AsyncSession,
    *,
    email: str,
    password: str,
    totpCode: str | None,
    expectedRole: str | None,
    ip: str | None,
    userAgent: str | None,
    background: BackgroundTasks,
) -> tuple[User, str, str]:
    normalized = _normalize_email(email)
    user = (await db.execute(
        select(User).where(func.lower(User.email) == normalized)
    )).scalar_one_or_none()

    # Generic-credential-error path (no user OR suspended/deleted account).
    is_active_account = (
        user is not None
        and (user.status is None or (user.status.value if hasattr(user.status, 'value') else str(user.status)).upper() in ("ACTIVE", "VERIFIED", "ENABLED", "OK"))
    )
    if not is_active_account or user is None:
        background.add_task(
            audit_service.record,
            action="auth.login.fail",
            ip=ip,
            userAgent=userAgent,
            metadata={"email": normalized, "reason": "unknown_user"},
        )
        raise AuthError(401, "Invalid credentials")

    locked_until_utc = _to_naive_utc(user.lockedUntil)
    if locked_until_utc is not None and locked_until_utc > _utcnow_naive():
        background.add_task(
            audit_service.record,
            actorId=user.id_,
            action="auth.login.locked",
            ip=ip,
            userAgent=userAgent,
        )
        raise AuthError(401, "Account temporarily locked. Try again later.")

    if not await security.async_verify_password(user.passwordHash, password):
        await _record_failed_attempt(db, user, ip=ip, userAgent=userAgent, background=background)
        raise AuthError(401, "Invalid credentials")

    if expectedRole and user.role != expectedRole:
        background.add_task(
            audit_service.record,
            actorId=user.id_,
            action="auth.login.wrong_portal",
            ip=ip,
            userAgent=userAgent,
        )
        raise AuthError(403, "Wrong sign-in portal for this account")

    # Success — reset counters, stamp last login (and upgrade legacy hash to Argon2id if needed).
    now = _utcnow_naive()
    update_values: dict[str, Any] = {
        "failedLoginCount": 0,
        "lockedUntil": None,
        "lastLoginAt": now,
        "updatedAt": now,
    }
    if security.needs_rehash(user.passwordHash):
        update_values["passwordHash"] = await security.async_hash_password(password)

    await db.execute(
        update(User)
        .where(User.id_ == user.id_)
        .values(**update_values)
    )

    # Re-load with the fresh values so the token reflects current state.
    fresh = (await db.execute(select(User).where(User.id_ == user.id_))).scalar_one()
    access_token = security.sign_access_token(_user_to_authed(fresh))
    refresh_token = await token_service.issue_refresh_token(
        db, fresh.id_, ip=ip, user_agent=userAgent
    )
    await db.commit()

    background.add_task(
        audit_service.record,
        actorId=user.id_,
        action="auth.login.success",
        ip=ip,
        userAgent=userAgent,
    )
    return fresh, access_token, refresh_token


async def _record_failed_attempt(
    db: AsyncSession,
    user: User,
    *,
    ip: str | None,
    userAgent: str | None,
    background: BackgroundTasks,
) -> None:
    s = get_settings()
    now = _utcnow_naive()
    window_start = now - timedelta(minutes=s.AUTH_LOCKOUT_WINDOW_MIN)
    user_locked_utc = _to_naive_utc(user.lockedUntil)
    user_updated_utc = _to_naive_utc(user.updatedAt)
    lockout_just_expired = user_locked_utc is not None and user_locked_utc < now
    counter_stale = user_updated_utc is not None and user_updated_utc < window_start
    should_reset = lockout_just_expired or counter_stale
    next_count = 1 if should_reset else user.failedLoginCount + 1
    should_lock = next_count >= s.AUTH_LOCKOUT_MAX_ATTEMPTS
    if should_lock:
        locked_until: datetime | None = now + timedelta(minutes=s.AUTH_LOCKOUT_DURATION_MIN)
    elif lockout_just_expired:
        locked_until = None
    else:
        locked_until = user.lockedUntil

    await db.execute(
        update(User)
        .where(User.id_ == user.id_)
        .values(failedLoginCount=next_count, lockedUntil=locked_until, updatedAt=now)
    )
    await db.commit()

    background.add_task(
        audit_service.record,
        actorId=user.id_,
        action="auth.login.lockout" if should_lock else "auth.login.fail",
        ip=ip,
        userAgent=userAgent,
        metadata={"failedCount": next_count},
    )


# ───────────── REFRESH / LOGOUT ──────────────────────────────────────────────


async def refresh(
    db: AsyncSession, *, presented_token: str, ip: str | None, userAgent: str | None
) -> tuple[str, str, User]:
    try:
        return await token_service.rotate(db, presented_token, ip=ip, user_agent=userAgent)
    except ValueError as e:
        raise AuthError(401, str(e))


async def logout(
    db: AsyncSession,
    *,
    presented_token: str | None,
    user_id: str | None,
    background: BackgroundTasks,
) -> None:
    if presented_token:
        await token_service.revoke_refresh_token(db, presented_token)
    if user_id:
        background.add_task(audit_service.record, actorId=user_id, action="auth.logout")


# ───────────── EMAIL VERIFICATION ────────────────────────────────────────────


async def verify_email(
    db: AsyncSession,
    *,
    token: str,
    ip: str | None,
    userAgent: str | None,
    background: BackgroundTasks,
) -> None:
    token_hash = _sha256(token)
    row = (await db.execute(
        select(EmailVerification).where(EmailVerification.tokenHash == token_hash)
    )).scalar_one_or_none()
    if row is None or row.usedAt is not None or row.expiresAt < _utcnow_naive():
        raise AuthError(400, "Invalid or expired verification token")
    now = _utcnow_naive()
    await db.execute(
        update(EmailVerification).where(EmailVerification.id_ == row.id_).values(usedAt=now)
    )
    await db.execute(
        update(User).where(User.id_ == row.userId).values(emailVerifiedAt=now, updatedAt=now)
    )
    await db.commit()
    background.add_task(
        audit_service.record,
        actorId=row.userId,
        action="auth.email.verified",
        ip=ip,
        userAgent=userAgent,
    )


async def resend_verification(db: AsyncSession, *, email: str) -> None:
    user = (await db.execute(
        select(User).where(User.email == _normalize_email(email))
    )).scalar_one_or_none()
    if user and user.emailVerifiedAt is None:
        raw = await _issue_verification_token(db, user_id=user.id_)
        await db.commit()
        # Best-effort send after commit — the token is already persisted.
        await _send_verification_email_safe(user.email, raw)
    # else: silent success to avoid enumeration


# ───────────── PASSWORD RESET ────────────────────────────────────────────────


async def forgot_password(
    db: AsyncSession,
    *,
    email: str,
    ip: str | None,
    userAgent: str | None,
    background: BackgroundTasks,
) -> None:
    s = get_settings()
    user = (await db.execute(
        select(User).where(User.email == _normalize_email(email))
    )).scalar_one_or_none()
    if user is None:
        return  # silent success

    raw = _nanoid(48)
    token_hash = _sha256(raw)
    now = _utcnow_naive()
    db.add(
        PasswordReset(
            id_=_mint_cuid(),
            userId=user.id_,
            tokenHash=token_hash,
            expiresAt=now + PASSWORD_RESET_TTL,
            createdAt=now,
        )
    )
    await db.flush()
    link = f"{s.WEB_ORIGIN}/reset-password?token={raw}"
    await get_mail_service().send_password_reset(user.email, link, totp_will_reset=user.totpEnabled)
    await db.commit()
    background.add_task(
        audit_service.record,
        actorId=user.id_,
        action="auth.password.reset_requested",
        ip=ip,
        userAgent=userAgent,
    )


async def reset_password(
    db: AsyncSession,
    *,
    token: str,
    newPassword: str,
    ip: str | None,
    userAgent: str | None,
    background: BackgroundTasks,
) -> bool:
    """Returns True if this reset also cleared an active TOTP enrollment."""
    token_hash = _sha256(token)
    row = (await db.execute(
        select(PasswordReset).where(PasswordReset.tokenHash == token_hash)
    )).scalar_one_or_none()
    if row is None or row.usedAt is not None or row.expiresAt < _utcnow_naive():
        raise AuthError(400, "Invalid or expired reset token")

    user = (await db.execute(select(User).where(User.id_ == row.userId))).scalar_one()
    had_totp = user.totpEnabled

    new_hash = await security.async_hash_password(newPassword)
    now = _utcnow_naive()
    await db.execute(
        update(PasswordReset).where(PasswordReset.id_ == row.id_).values(usedAt=now)
    )
    await db.execute(
        update(User)
        .where(User.id_ == row.userId)
        .values(
            passwordHash=new_hash,
            failedLoginCount=0,
            lockedUntil=None,
            # Proving email ownership is also our TOTP-recovery path: an admin who
            # lost their authenticator can't reach /2fa/setup without first logging
            # in, and login requires the (now-missing) code. Clearing it here lets
            # them log in and re-enroll via /admin/setup-2fa.
            totpEnabled=False,
            totpSecret=None,
            updatedAt=now,
        )
    )
    # Invalidate all sessions.
    await db.execute(
        update(RefreshToken)
        .where(RefreshToken.userId == row.userId, RefreshToken.revokedAt.is_(None))
        .values(revokedAt=now)
    )
    await db.commit()
    background.add_task(
        audit_service.record,
        actorId=row.userId,
        action="auth.password.reset",
        ip=ip,
        userAgent=userAgent,
        metadata={"totpReset": had_totp},
    )
    return had_totp


# ───────────── 2FA ───────────────────────────────────────────────────────────


async def begin_two_factor_setup(db: AsyncSession, *, user_id: str) -> tuple[str, str]:
    """Returns (otpauthUrl, qrCodeDataUrl). Stores unconfirmed secret on the user row."""
    try:
        import qrcode  # type: ignore[import]
    except ImportError as exc:
        raise RuntimeError("The qrcode package is required for 2FA setup") from exc
    import io
    import base64

    user = (await db.execute(select(User).where(User.id_ == user_id))).scalar_one()
    secret_b32 = security.generate_totp_secret()
    otpauth = security.totp_uri(user.email, secret_b32)

    # QR as data URL — match Nest's qrcode.toDataURL() output shape.
    img = qrcode.make(otpauth)
    buf = io.BytesIO()
    img.save(buf)
    qr_data_url = "data:image/png;base64," + base64.b64encode(buf.getvalue()).decode("ascii")

    await db.execute(
        update(User)
        .where(User.id_ == user_id)
        .values(totpSecret=secret_b32, totpEnabled=False, updatedAt=_utcnow_naive())
    )
    await db.commit()
    return otpauth, qr_data_url


async def confirm_two_factor_setup(
    db: AsyncSession,
    *,
    user_id: str,
    code: str,
    ip: str | None,
    userAgent: str | None,
    background: BackgroundTasks,
) -> None:
    user = (await db.execute(select(User).where(User.id_ == user_id))).scalar_one()
    if not user.totpSecret:
        raise AuthError(400, "No 2FA setup in progress")
    if not security.verify_totp(code, user.totpSecret):
        raise AuthError(400, "Invalid 2FA code")
    await db.execute(
        update(User)
        .where(User.id_ == user_id)
        .values(totpEnabled=True, updatedAt=_utcnow_naive())
    )
    await db.commit()
    background.add_task(
        audit_service.record,
        actorId=user_id,
        action="auth.2fa.enabled",
        ip=ip,
        userAgent=userAgent,
    )


async def disable_two_factor(
    db: AsyncSession,
    *,
    user_id: str,
    password: str,
    code: str,
    ip: str | None,
    userAgent: str | None,
    background: BackgroundTasks,
) -> None:
    user = (await db.execute(select(User).where(User.id_ == user_id))).scalar_one()
    if user.role == "ADMIN":
        raise AuthError(403, "Admins cannot disable 2FA")
    if not await security.async_verify_password(user.passwordHash, password):
        raise AuthError(401, "Invalid password")
    if not user.totpEnabled or not user.totpSecret:
        raise AuthError(400, "2FA is not enabled")
    if not security.verify_totp(code, user.totpSecret):
        raise AuthError(400, "Invalid 2FA code")
    await db.execute(
        update(User)
        .where(User.id_ == user_id)
        .values(totpEnabled=False, totpSecret=None, updatedAt=_utcnow_naive())
    )
    await db.commit()
    background.add_task(
        audit_service.record,
        actorId=user_id,
        action="auth.2fa.disabled",
        ip=ip,
        userAgent=userAgent,
    )


# ───────────── ME ────────────────────────────────────────────────────────────


async def me(db: AsyncSession, *, user_id: str) -> dict[str, Any]:
    user = (await db.execute(
        select(User).options(joinedload(User.profile)).where(User.id_ == user_id)
    )).scalar_one()
    profile = user.profile
    return {
        "id": user.id_,
        "email": user.email,
        "role": user.role,
        "status": user.status,
        "emailVerified": user.emailVerifiedAt is not None,
        "totpEnabled": user.totpEnabled,
        "profile": (
            {
                "id": profile.id_,
                "userId": profile.userId,
                "firstName": profile.firstName,
                "lastName": profile.lastName,
                "phone": profile.phone,
                "dob": _iso_ms(profile.dob) if profile.dob else None,
                "avatarUrl": profile.avatarUrl,
                "tier": profile.tier,
                "createdAt": _iso_ms(profile.createdAt),
                "updatedAt": _iso_ms(profile.updatedAt),
            }
            if profile
            else None
        ),
        "createdAt": _iso_ms(user.createdAt),
    }


def _iso_ms(dt: datetime) -> str:
    """ISO 8601 with millisecond precision + 'Z' suffix — matches Nest's
    `new Date().toISOString()` exactly. Python's `.isoformat()` would emit
    microseconds (`.319000`), which would cause a one-character parity diff."""
    # Strip to milliseconds and append Z.
    return dt.strftime("%Y-%m-%dT%H:%M:%S.") + f"{dt.microsecond // 1000:03d}Z"


# ───────────── helpers ───────────────────────────────────────────────────────


def _user_to_authed(u: User) -> AuthenticatedUser:
    return AuthenticatedUser(
        id=u.id_,
        email=u.email,
        role=u.role,  # type: ignore[arg-type]
        totpEnabled=u.totpEnabled,
        emailVerified=u.emailVerifiedAt is not None,
    )
